strawberryfields/io/xir_io.py

```diff
@@ -0,0 +1,294 @@
+                    gate(**params) | regrefs  # pylint:disable=expression-not-assigned
+                else:
+                    params = []
+                    for param in op.params:
+                        if isinstance(param, Decimal):
+                            params.append(float(param))
+                        elif isinstance(param, (list, np.ndarray)):
+                            params.append(np.array(_listr(param)))
+                        elif isinstance(param, str) and param[0] == "p" and param[1:].isdigit():
+                            params.append(p[int(param[1:])])
+                        else:
+                            params.append(param)
+                    params = sfpar.par_convert(params, prog)
+                    gate(*params) | regrefs  # pylint:disable=expression-not-assigned
+            else:
+                if callable(gate):
+                    gate() | regrefs  # pylint:disable=expression-not-assigned,pointless-statement
+                else:
+                    gate | regrefs  # pylint:disable=expression-not-assigned,pointless-statement
+
+    prog._target = xir_prog.options.get("target", None)  # pylint: disable=protected-access
+
+    if "shots" in xir_prog.options:
+        prog.run_options["shots"] = xir_prog.options["shots"]
+    if "cutoff_dim" in xir_prog.options:
+        prog.backend_options["cutoff_dim"] = xir_prog.options["cutoff_dim"]
+
+    return prog
+
+
+def to_xir(prog, **kwargs):
+    """Convert a Strawberry Fields Program to an XIR Program.
+
+    Args:
+        prog (Program): the Strawberry Fields program
+
+    Keyword Args:
+        add_decl (bool): Whether gate and output declarations should be added to
+            the XIR program. Default is ``False``.
+        version (str): Version number for the program. Default is ``0.1.0``.
+
+    Returns:
+        xir.Program
+    """
+    version = kwargs.get("version", "0.1.0")
+    xir_prog = xir.Program(version=version)
+
+    if isinstance(prog, TDMProgram):
+        xir_prog.add_option("type", "tdm")
+        xir_prog.add_option("N", prog.N)
+        for i, p in enumerate(prog.tdm_params):
+            xir_prog.add_constant(f"p{i}", _listr(p))
+
+    if prog._target:
+        xir_prog.add_option("target", prog._target)  # pylint: disable=protected-access
+    if "cutoff_dim" in prog.run_options:
+        xir_prog.add_option("cutoff_dim", prog.run_options["cutoff_dim"])
+    if "name" in prog.run_options:
+        xir_prog.add_option("name", prog.run_options["name"])
+    if "shots" in prog.run_options:
+        xir_prog.add_option("shots", prog.run_options["shots"])
+
+    # fill in the quantum circuit
+    for cmd in prog.circuit or []:
+
+        name = cmd.op.__class__.__name__
+        wires = tuple(i.ind for i in cmd.reg)
+
+        if "Measure" in name:
+            if kwargs.get("add_decl", False):
+                output_decl = xir.Declaration(name, type_="out")
```

Comment:
Should there also be `...` in place of the wires?

Reply:
`...` isn't supported yet, so I don't think we should add it just yet, but the wires should definitely be added.